content/blog/2022/zero-units.md

```diff
@@ -0,0 +1,359 @@
+
+In some cases, `1` can even be a `<string>`.
+While [CSS counters](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Counter_Styles/Using_CSS_counters)
+are clearly counting with numbers,
+the output from `counter()` and `counters()`
+will always be a `<string>` representation of that count.
+For now, the same is true for output of the `attr()` function.
+That's part of why we can't (currently)
+use counters & attributes to do much outside generated content.
+(The other reason is that these functions only work in the `content` property,
+but the logic of that is a bit recursive --
+if the only output is a `<string>`,
+and only `content` accepts `<string>` values,
+there's no reason to allow counters anywhere else.)
+
+And CSS doesn't generally allow _coercing_ values
+from one type to another.
+There's no way to take a string and turn it into a number,
+or vice versa.
+We can convert a number into a length (or time) --
+`calc(<number> * <length>)` will return a `<length>` value --
+but we can't (yet) go the other way:[^unit-division]
+
+```css
+.example {
+  --number: 3;
+  /* converts the number 3 to the length of 3em */
+  margin-block: calc(var(--number) * 1em);
+}
+```
+
+[^design-types]: This is one reason
+  I strongly prefer using [Sass](https://sass-lang.com/)
+  to manage 'tokens' in a design system,
+  rather than a language like YAML or JSON.
+  Sass is entirely organized around the CSS type system
+  and 'design-relevant' types --
+  lengths, colors, etc.
+  Languages that are not intended for 'design' specifically
+  tend to have very different value types.
+
+[^unit-division]: The spec allows for removing units through division,
+  but no browsers have implemented that feature yet.
+
+## Zero is (often) special
+
+In most cases, zero is an exception to the type rules --
+we can use it in many places as either a `<number>` or a `<length>`
+without adding any units!
+That's because `0` is the same length (no length!),
+no matter what units you apply to it.
+Zero `em` is the same as zero `px` and zero `%` and so on.
+You can't set `margin` to `5` (a `<number>`),
+but you _can_ set it to `0` (also a `<number>`).
+
+For _zero and only zero_,
+we can use a `<number>` when CSS expects a `<length>`.
+
+And over time,
+that has become a 'best practice' --
+often enforced by CSS linters & minifiers.
+The usual reasoning is performance.
+Removing all the units from zeros
+will save you a couple bytes
+for every occurrence.
+You could also consider it better for readability --
+if all zero values are equal,
+units only distract from the meaning.
+
+## Zero is (not always) special
+
+That 'best practice' works great
+for raw zero values,
+directly applied to properties like
+`margin` or `padding` --
+but there are other places
+where this 'best practice' will break your CSS.
+
+In general:
+_when zero is inside a function, the type of zero matters_.
+(At least,
+that's where I've always encountered the issue.)
```

Comment:
It might be interesting to close the article with a audience call to action, like what cases have you come across ...   or maybe as part of the tweet that shares this @mirisuzanne